File: sdk-go/dex/flow.go
// Legacy Materials in this file remain under their original licenses.
// See LEGACY_NOTICES.md.

// Modifications Copyright (c) 2026 Super Durable, Inc.
//
// Modifications after the Legacy Cutoff are licensed under the
// Super Durable Source License 1.0.
// Legacy Materials remain under their original licenses.
// See LICENSE and LEGACY_NOTICES.md.

package dex

// Flow is the top-level durable workflow definition. Any long-lived business
// object (at least a few seconds) can be modeled as a Flow: a Go type, a list
// of Steps, and a persistence schema of attributes and channels.
//
// Exported methods on the Flow (other than the Flow interface methods) that
// match RPC[IN, OUT] are registered as RPCs under their Go method names.
//
// Example:
//
//	type OrderFlow struct {
//		dex.FlowDefaults
//	}
//
//	func (OrderFlow) GetSteps() []dex.StepDef {
//		return []dex.StepDef{
//			dex.DefineStartStep(ApproveOrder),
//			dex.DefineStep(ShipOrder),
//		}
//	}
//
//	func (OrderFlow) GetPersistenceSchema() dex.PersistenceSchema {
//		return dex.PersistenceSchema{
//			Attributes: []dex.AttributeDef{OrderStatus},
//			Channels:   []dex.ChannelDef{ApprovalChannel},
//		}
//	}
//
//	func (OrderFlow) GetSnapshot(
//		ctx dex.Context,
//		input GetSnapshotInput,
//	) (*dex.RPCResult[OrderSnapshot], error) {
//		return &dex.RPCResult[OrderSnapshot]{Output: OrderSnapshot{}}, nil
//	}
//
//	var Orders = OrderFlow{}
//	var _ dex.Flow = Orders
type Flow interface {
	// GetFlowType overrides the default package-qualified Go type name.
	// Embed FlowDefaults to use the default.
	GetFlowType() string

	// GetSteps defines the steps of the flow. A step is one node in the flow
	// state machine: it may WaitFor conditions (channel / timer) and then
	// Execute a decision. See Step for details.
	//
	// Use DefineStartStep for at most one starting step and DefineStep for
	// every other step. An empty list, or a list with no starting step, means
	// the run starts with no step execution; application code can still invoke
	// RPCs that move into steps later.
	GetSteps() []StepDef

	// GetPersistenceSchema declares attributes and channels for this flow.
	//
	// Attributes can be read and written in WaitFor, Execute, and RPC. External
	// clients can also read them through Client APIs. Indexed attributes can be
	// used in SearchFlows queries.
	//
	// Channels carry messages into a waiting step: external clients publish;
	// WaitFor requests consumption and Execute reads results via
	// Channel.GetConditionResults. Channel maps key messages by string.
	GetPersistenceSchema() PersistenceSchema
}

// FlowTimeoutHandler lets a Flow execute application code when its soft timeout expires.
//
// Implement this capability on the same Flow value registered with NewRegistry. A positive
// StartFlowOptions.Timeout defaults to TimeoutHandler when this capability is present. The
// returned StepDecision has the same validation and terminal behavior as a Step Execute result.
// Context exposes the timed-out Flow execution and permits normal Attribute and Channel writes.
//
// HandleTimeout may return GoTo, DeadEnd, ForceComplete, ForceFail, or GracefulComplete. It runs
// at most once for an execution, after Dex's durable timeout timer completes or is skipped.
type FlowTimeoutHandler interface {
	// HandleTimeout executes the Flow's soft-timeout hook.
	//
	// ctx belongs to this invocation and must not be retained. A non-nil error applies normal
	// Execute retry and failure policy behavior; a nil decision is rejected as invalid.
	HandleTimeout(ctx Context) (*StepDecision, error)
}

// FlowDefaults uses the package-qualified Go type as the durable flow type.
type FlowDefaults struct{}

// GetFlowType returns empty so Registry derives the package-qualified Go type name.
func (FlowDefaults) GetFlowType() string {
	return ""
}

// PersistenceSchema is the set of AttributeDef and ChannelDef values a Flow
// registers. Every attribute or channel used from WaitFor, Execute, or RPC must
// appear here.
type PersistenceSchema struct {
	// Attributes contains every Attribute and AttributeMap used by the Flow.
	Attributes []AttributeDef
	// Channels contains every Channel and ChannelMap used by the Flow.
	Channels []ChannelDef
}

File: sdk-java/src/main/java/io/superdurable/dex/Flow.java
/*
 * Legacy Materials in this file remain under their original licenses.
 * See LEGACY_NOTICES.md.
 */

/*
 * Modifications Copyright (c) 2026 Super Durable, Inc.
 *
 * Modifications after the Legacy Cutoff are licensed under the
 * Super Durable Source License 1.0.
 * Legacy Materials remain under their original licenses.
 * See LICENSE and LEGACY_NOTICES.md.
 */

package io.superdurable.dex;

/**
 * Defines the Steps, persistence, and RPC surface of one Flow type.
 *
 * <p>Applications implement this interface on a concrete class and register an instance with
 * {@link Registry}. The generic parameter is the input accepted by the start Step; the typed
 * {@link StepList#startStep} factory enforces that relationship at compile time. A Flow may omit a
 * start Step and begin only through an RPC. The default Flow type is the concrete class's simple
 * name, so explicit named classes make that type predictable and reviewable.
 *
 * <pre>{@code
 * final class OrderFlow implements Flow<OrderInput> {
 *     private final ValidateOrder start = new ValidateOrder();
 *     private final ChargeOrder charge = new ChargeOrder();
 *
 *     @Override
 *     public StepList<OrderInput> getSteps() {
 *         return StepList.startStep(start).otherSteps(charge);
 *     }
 * }
 * }</pre>
 *
 * @param <StartInput> the Java input type accepted by the Flow's start Step
 */
public interface Flow<StartInput> {
    /**
     * Returns the Flow type used to register and start this Flow.
     *
     * <p>The default is {@code getClass().getSimpleName()}. Override it when refactoring the Java
     * class name must not change the Flow type stored by Dex.
     *
     * @return the nonblank Flow type
     */
    default String getFlowType() {
        return getClass().getSimpleName();
    }

    /**
     * Returns the Flow's start Step and all other registered Steps.
     *
     * <p>The default is an empty list, which is valid for an RPC-only Flow. Every Step targeted by a
     * decision or RPC result must appear in this list.
     *
     * @return the immutable typed Step list; never {@code null}
     */
    default StepList<StartInput> getSteps() {
        return StepList.empty();
    }

    /**
     * Returns the persistent Attributes and Channels used by this Flow.
     *
     * <p>The default schema is empty. Every persistence definition used by Steps or RPC methods must
     * be registered here.
     *
     * @return the Flow's immutable persistence schema; never {@code null}
     */
    default PersistenceSchema getPersistenceSchema() {
        return PersistenceSchema.of();
    }

    /**
     * Handles expiration of this Flow's soft timeout.
     *
     * <p>Override this method to make a positive timeout default to
     * {@link FlowTimeoutPolicy#HANDLER}. Dex calls it at most once after the durable timeout timer
     * completes or is skipped. The returned decision may transition to another Step, end without
     * closing, complete, fail, or request graceful completion. The Context belongs to this
     * invocation and must not be retained.
     *
     * @param context the timeout-handler invocation Context
     * @return the non-null decision applied with normal Step Execute semantics
     * @throws UnsupportedOperationException when called without an application override
     */
    default StepDecision handleTimeout(final Context context) {
        throw new UnsupportedOperationException("Flow has no timeout handler");
    }
}
